src/crypto/CHIPCryptoPALOpenSSL.cpp

```diff
@@ -1703,6 +1703,54 @@ CHIP_ERROR ValidateCertificateChain(const uint8_t * rootCertificate, size_t root
     return err;
 }
 
+CHIP_ERROR IsCertificateTimestampValid(CertificateValidityType validityType, const ByteSpan & referenceCertificate,
+                                       const ByteSpan & toBeEvaluatedCertificate)
+{
+    CHIP_ERROR error                                = CHIP_NO_ERROR;
+    X509 * x509ReferenceCertificate                 = nullptr;
+    X509 * x509toBeEvaluatedCertificate             = nullptr;
+    const unsigned char * pReferenceCertificate     = referenceCertificate.data();
+    const unsigned char * pToBeEvaluatedCertificate = toBeEvaluatedCertificate.data();
+    ASN1_TIME * timeA                               = nullptr;
+    ASN1_TIME * timeB                               = nullptr;
+    int result                                      = 0;
+
+    VerifyOrReturnError(!referenceCertificate.empty() && !toBeEvaluatedCertificate.empty(), CHIP_ERROR_INVALID_ARGUMENT);
+
+    x509ReferenceCertificate = d2i_X509(NULL, &pReferenceCertificate, static_cast<long>(referenceCertificate.size()));
+    VerifyOrExit(x509ReferenceCertificate != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+    x509toBeEvaluatedCertificate = d2i_X509(NULL, &pToBeEvaluatedCertificate, static_cast<long>(toBeEvaluatedCertificate.size()));
+    VerifyOrExit(x509toBeEvaluatedCertificate != nullptr, error = CHIP_ERROR_NO_MEMORY);
+
+    switch (validityType)
+    {
+    case CertificateValidityType::kNotAfter:
+        timeB = X509_get_notAfter(x509ReferenceCertificate);
+        timeA = X509_get_notAfter(x509toBeEvaluatedCertificate);
+        break;
+    case CertificateValidityType::kNotBefore:
+        timeA = X509_get_notBefore(x509ReferenceCertificate);
+        timeB = X509_get_notBefore(x509toBeEvaluatedCertificate);
+        break;
+    default:
+        SuccessOrExit(error = CHIP_ERROR_INVALID_ARGUMENT);
+    }
+
+    VerifyOrExit(timeA && timeB, error = CHIP_ERROR_INTERNAL);
+
+    result = ASN1_TIME_compare(timeA, timeB);
+    // check if tbeCertificate is valid before/during referenceCertificate's notBefore timestamp.
+    // check if tbeCertificate is valid after/during referenceCertificate's notAfter timestamp.
+    VerifyOrExit(result >= 0, error = CHIP_ERROR_CERT_EXPIRED);
```

Comment:
This check is more confusing than it needs to be.

Suggest splitting in 2 separate blocks, one for notBefore, one for notAfter, rather than attempting to use common code for both.

The check needed is the following pseudocode:

```
bool validAtIssuance = (dac.notBefore >= ref.notBefore) && (dac.notBefore <= ref.notAfter);
```

Reply:
IsCertificateTimestampValid() method is currently being called twice (for NotBefore and NotAfter). Are you suggesting to remove CertificateValidityType param and combine the both the time validity checks?

Reply:
Not quite, although having 1 method with all validity checks would be simpler to understand.

The split of TimeA/TimeB and their reversal cases makes it hard to understand how the comparison works. Reading the code top to bottom I am struggling to make sure everything is considered, especially since Time A and Time B reverse roles.

The check `validAtIssuance` relates to the DAC being valid at issuance time according to Device Attestation Procedure rules. The DAC overall validity check is a separate check.

Reply:
Fixed in 4ef7efd